File: crates/worker/js/runtime/persona.js
// Installs deterministic API-shape emulation from a fully resolved persona.
// Rendering, media processing, device access, and network services stay native
// or absent; this module only owns values that can be represented in JavaScript.
function installPersona(persona, runtimeFeatures) {
    "use strict";

    const navigatorValues = persona.navigator;
    const network = persona.base_headers;
    const markBuiltin = globalThis.__brimpMarkWebBuiltin;
    const markNative = (fn, name = undefined) => {
        if (typeof fn === "function") {
            if (name === undefined) markBuiltin(fn);
            else markBuiltin(fn, `function ${name}() { [native code] }`);
        }
        return fn;
    };
    const markAccessor = (fn, kind, name) => {
        Object.defineProperty(fn, "name", { value: `${kind} ${name}`, configurable: true });
        markBuiltin(fn, `function ${kind} ${name}() { [native code] }`);
        return fn;
    };
    const markInterface = constructor => {
        markNative(constructor);
        if (!constructor.prototype) return constructor;
        if (!Object.hasOwn(constructor.prototype, Symbol.toStringTag)) {
            Object.defineProperty(constructor.prototype, Symbol.toStringTag, {
                value: constructor.name,
                writable: false,
                enumerable: false,
                configurable: true,
            });
        }
        for (const key of Reflect.ownKeys(constructor.prototype)) {
            const descriptor = Object.getOwnPropertyDescriptor(constructor.prototype, key);
            if (typeof descriptor?.value === "function") markNative(descriptor.value);
            if (typeof descriptor?.get === "function") markAccessor(descriptor.get, "get", String(key));
            if (typeof descriptor?.set === "function") markAccessor(descriptor.set, "set", String(key));
        }
        return constructor;
    };
    const markedObjects = new WeakSet();
    const markObjectFunctions = value => {
        if ((typeof value !== "object" && typeof value !== "function") || value === null) return;
        if (markedObjects.has(value)) return;
        markedObjects.add(value);
        for (const key of Reflect.ownKeys(value)) {
            const descriptor = Object.getOwnPropertyDescriptor(value, key);
            if (typeof descriptor?.value === "function") markNative(descriptor.value, String(key));
            else if (typeof descriptor?.value === "object") markObjectFunctions(descriptor.value);
            if (typeof descriptor?.get === "function") markAccessor(descriptor.get, "get", String(key));
            if (typeof descriptor?.set === "function") markAccessor(descriptor.set, "set", String(key));
        }
    };
    const defineValue = (target, name, value, enumerable = true) => {
        if (typeof value === "function") markInterface(value);
        else markObjectFunctions(value);
        Object.defineProperty(target, name, {
            value,
            enumerable,
            configurable: true,
        });
    };
    const defineGlobalConstructor = (name, constructor) => {
        markInterface(constructor);
        Object.defineProperty(globalThis, name, {
            value: constructor,
            writable: true,
            enumerable: false,
            configurable: true,
        });
    };
    const navigatorState = Object.create(null);
    const defineNavigatorValue = (name, value) => {
        navigatorState[name] = value;
        markObjectFunctions(value);
        const getter = markAccessor(function () { return navigatorState[name]; }, "get", name);
        Object.defineProperty(Navigator.prototype, name, {
            get: getter,
            enumerable: true,
            configurable: true,
        });
    };

    for (const [name, value] of Object.entries({
        userAgent: network.user_agent,
        appVersion: network.user_agent.replace(/^Mozilla\//, ""),
        appCodeName: "Mozilla",
        appName: "Netscape",
        product: "Gecko",
        platform: navigatorValues.platform,
        language: persona.locale,
        languages: Object.freeze(persona.languages.slice()),
        hardwareConcurrency: navigatorValues.hardware_concurrency,
        maxTouchPoints: navigatorValues.max_touch_points,
        doNotTrack: navigatorValues.do_not_track || null,
        vendor: navigatorValues.vendor,
        productSub: navigatorValues.product_sub,
        pdfViewerEnabled: navigatorValues.pdf_viewer_enabled,
        cookieEnabled: true,
        onLine: true,
    })) defineNavigatorValue(name, value);
    defineNavigatorValue("deviceMemory", navigatorValues.device_memory_gb);
    if (navigatorValues.expose_global_privacy_control) {
        defineNavigatorValue("globalPrivacyControl", navigatorValues.global_privacy_control);
    }

    const constructToken = Symbol("browser interface construction");
    class NetworkInformation extends EventTarget {
        constructor(...args) {
            super();
            if (args[0] !== constructToken) throw new TypeError("Illegal constructor");
        }
        get type() { return navigatorValues.connection_type; }
        get rtt() { return navigatorValues.connection_rtt_ms; }
        get downlink() { return Number(navigatorValues.connection_downlink_mbps); }
        get effectiveType() { return navigatorValues.connection_effective_type; }
        get saveData() { return navigatorValues.connection_save_data; }
    }
    defineNavigatorValue("connection", new NetworkInformation(constructToken));

    const parseBrands = header => {
        const result = [];
        const pattern = /"([^"]+)";v="([^"]+)"/g;
        for (let match; (match = pattern.exec(header)) !== null;) {
            result.push(Object.freeze({ brand: match[1], version: match[2] }));
        }
        return Object.freeze(result);
    };
    const unquote = value => value.replace(/^"|"$/g, "");
    if (network.sec_ch_ua) {
        const brands = parseBrands(network.sec_ch_ua);
        const fullVersionList = parseBrands(network.sec_ch_ua_full_version_list);
        class NavigatorUAData {
            constructor(...args) {
                if (args[0] !== constructToken) throw new TypeError("Illegal constructor");
            }
            get brands() { return brands; }
            get mobile() { return network.sec_ch_ua_mobile === "?1"; }
            get platform() { return unquote(network.sec_ch_ua_platform); }
            getHighEntropyValues(hints) {
                const values = {
                    architecture: unquote(network.sec_ch_ua_arch),
                    bitness: unquote(network.sec_ch_ua_bitness),
                    model: unquote(network.sec_ch_ua_model),
                    platformVersion: unquote(network.sec_ch_ua_platform_version),
                    uaFullVersion: unquote(network.sec_ch_ua_full_version),
                    fullVersionList,
                    wow64: false,
                };
                const selected = this.toJSON();
                for (const hint of hints) {
                    if (Object.hasOwn(values, hint)) selected[hint] = values[hint];
                }
                return Promise.resolve(selected);
            }
            toJSON() {
                return { brands, mobile: this.mobile, platform: this.platform };
            }
        }
        defineGlobalConstructor("NavigatorUAData", NavigatorUAData);
        defineNavigatorValue("userAgentData", new NavigatorUAData(constructToken));
    }

    const screenValues = persona.screen;
    class ScreenOrientation extends EventTarget {
        constructor(...args) {
            super();
            if (args[0] !== constructToken) throw new TypeError("Illegal constructor");
        }
        get type() { return screenValues.orientation_type; }
        get angle() { return screenValues.orientation_angle; }
    }
    const orientation = new ScreenOrientation(constructToken);
    class Screen {
        constructor() { throw new TypeError("Illegal constructor"); }
        get width() { return screenValues.width; }
        get height() { return screenValues.height; }
        get availWidth() { return screenValues.avail_width; }
        get availHeight() { return screenValues.avail_height; }
        get availLeft() { return screenValues.avail_left; }
        get availTop() { return screenValues.avail_top; }
        get left() { return screenValues.left; }
        get top() { return screenValues.top; }
        get colorDepth() { return screenValues.color_depth; }
        get pixelDepth() { return screenValues.pixel_depth; }
        get isExtended() { return screenValues.is_extended; }
        get orientation() { return orientation; }
    }
    const screenObject = Object.create(Screen.prototype);
    defineGlobalConstructor("Screen", Screen);
    defineGlobalConstructor("ScreenOrientation", ScreenOrientation);
    defineValue(globalThis, "screen", screenObject, false);

    const windowValues = persona.window;
    Object.defineProperties(globalThis, {
        outerWidth: { value: windowValues.outer_width, configurable: true },
        outerHeight: { value: windowValues.outer_height, configurable: true },
        screenX: { value: windowValues.screen_x, configurable: true },
        screenY: { value: windowValues.screen_y, configurable: true },
        screenLeft: { value: windowValues.screen_x, configurable: true },
        screenTop: { value: windowValues.screen_y, configurable: true },
    });

    const arrayData = new WeakMap();
    class MimeType {
        constructor() { throw new TypeError("Illegal constructor"); }
        get type() { return arrayData.get(this).type; }
        get suffixes() { return arrayData.get(this).suffixes; }
        get description() { return arrayData.get(this).description; }
        get enabledPlugin() { return arrayData.get(this).enabledPlugin; }
    }
    class Plugin {
        constructor() { throw new TypeError("Illegal constructor"); }
        get name() { return arrayData.get(this).name; }
        get filename() { return arrayData.get(this).filename; }
        get description() { return arrayData.get(this).description; }
        get length() { return arrayData.get(this).items.length; }
        item(index) { return arrayData.get(this).items[Number(index)] ?? null; }
        namedItem(name) { return arrayData.get(this).names[String(name)] ?? null; }
        [Symbol.iterator]() { return arrayData.get(this).items[Symbol.iterator](); }
    }
    class PluginArray {
        constructor() { throw new TypeError("Illegal constructor"); }
        get length() { return arrayData.get(this).items.length; }
        item(index) { return arrayData.get(this).items[Number(index)] ?? null; }
        namedItem(name) { return arrayData.get(this).names[String(name)] ?? null; }
        refresh() {}
        [Symbol.iterator]() { return arrayData.get(this).items[Symbol.iterator](); }
    }
    class MimeTypeArray {
        constructor() { throw new TypeError("Illegal constructor"); }
        get length() { return arrayData.get(this).items.length; }
        item(index) { return arrayData.get(this).items[Number(index)] ?? null; }
        namedItem(name) { return arrayData.get(this).names[String(name)] ?? null; }
        [Symbol.iterator]() { return arrayData.get(this).items[Symbol.iterator](); }
    }
    const makeArrayLike = (constructor, items, names) => {
        const value = Object.create(constructor.prototype);
        arrayData.set(value, { items, names });
        items.forEach((item, index) => defineValue(value, index, item));
        for (const [name, item] of Object.entries(names)) {
            if (!(name in value)) defineValue(value, name, item, false);
        }
        return value;
    };
    const mimeByType = Object.create(null);
    const pluginByName = Object.create(null);
    const pdfPlugins = navigatorValues.pdf_viewer_enabled ? [
        "PDF Viewer", "Chrome PDF Viewer", "Chromium PDF Viewer",
        "Microsoft Edge PDF Viewer", "WebKit built-in PDF",
    ] : [];
    const pluginEntries = pdfPlugins.map(name => {
        const pluginConfig = {
            name,
            filename: "internal-pdf-viewer",
            description: "Portable Document Format",
            mime_types: ["application/pdf", "text/pdf"].map(type => ({
                type, suffixes: "pdf", description: "Portable Document Format",
            })),
        };
        const plugin = Object.create(Plugin.prototype);
        const mimeNames = Object.create(null);
        const mimeEntries = pluginConfig.mime_types.map(mime => {
            const value = Object.create(MimeType.prototype);
            arrayData.set(value, { ...mime, enabledPlugin: plugin });
            mimeNames[mime.type] = value;
            if (!(mime.type in mimeByType)) mimeByType[mime.type] = value;
            return value;
        });
        arrayData.set(plugin, {
            name: pluginConfig.name,
            filename: pluginConfig.filename,
            description: pluginConfig.description,
            items: mimeEntries,
            names: mimeNames,
        });
        mimeEntries.forEach((item, index) => defineValue(plugin, index, item));
        pluginByName[pluginConfig.name] = plugin;
        return plugin;
    });
    const plugins = makeArrayLike(PluginArray, pluginEntries, pluginByName);
    const mimeTypes = makeArrayLike(MimeTypeArray, Object.values(mimeByType), mimeByType);
    for (const constructor of [MimeType, Plugin, PluginArray, MimeTypeArray]) {
        defineGlobalConstructor(constructor.name, constructor);
    }
    defineNavigatorValue("plugins", plugins);
    defineNavigatorValue("mimeTypes", mimeTypes);

    {
        function Notification(title, options = {}) {
            this.title = String(title);
            this.body = options.body === undefined ? "" : String(options.body);
            this.close = () => undefined;
        }
        Object.defineProperty(Notification, "permission", {
            get: () => navigatorValues.notification_permission,
        });
        Notification.requestPermission = callback => {
            const permission = navigatorValues.notification_permission;
            if (typeof callback === "function") callback(permission);
            return Promise.resolve(permission);
        };
        defineGlobalConstructor("Notification", Notification);
    }

    if (navigatorValues.permissions_enabled) {
        const permissions = Object.freeze({
            query(descriptor) {
                const state = descriptor && descriptor.name === "notifications"
                    ? (navigatorValues.notification_permission === "default" ? "prompt" : navigatorValues.notification_permission)
                    : "prompt";
                return Promise.resolve(Object.freeze({
                    state,
                    onchange: null,
                    addEventListener() {},
                    removeEventListener() {},
                }));
            },
        });
        defineNavigatorValue("permissions", permissions);
    } else {
        delete Navigator.prototype.permissions;
    }

    if (navigatorValues.bluetooth_enabled) {
        defineNavigatorValue("bluetooth", Object.freeze({
            getAvailability: () => Promise.resolve(navigatorValues.bluetooth_available),
        }));
    } else {
        delete Navigator.prototype.bluetooth;
    }

    if (navigatorValues.media_devices_enabled) {
        defineNavigatorValue("mediaDevices", Object.freeze({
            enumerateDevices: () => Promise.resolve([]),
            getSupportedConstraints: () => Object.freeze({}),
        }));
    } else {
        delete Navigator.prototype.mediaDevices;
    }

    if (!navigatorValues.offscreen_canvas_enabled) {
        delete globalThis.OffscreenCanvas;
        delete globalThis.OffscreenCanvasRenderingContext2D;
        delete HTMLCanvasElement.prototype.transferControlToOffscreen;
    }

    if (!runtimeFeatures.persistentStorage) {
        delete Navigator.prototype.storage;
        delete Navigator.prototype.webkitTemporaryStorage;
        delete Navigator.prototype.webkitPersistentStorage;
    }

    const applyIdentityOverride = serialized => {
        const override = JSON.parse(serialized);
        for (const name of ["userAgent", "platform", "language", "languages"]) {
            if (Object.hasOwn(override, name) && override[name] !== null) {
                navigatorState[name] = name === "languages"
                    ? Object.freeze(Array.from(override[name], String))
                    : String(override[name]);
            }
        }
        return true;
    };
    globalThis.__brimpFinalizeWebIdl();
    delete globalThis.__brimpFinalizeWebIdl;
    delete globalThis.__brimpMarkWebBuiltin;
    return applyIdentityOverride;
}

File: crates/worker/tests/runtime/persona.rs
use async_trait::async_trait;
use brimp_lite_worker::network::{NetworkError, ResourceLoader, ResourceRequest, ResourceResponse};
use brimp_lite_worker::runtime::{Browser, PageOptions};
use http::{HeaderMap, HeaderValue, StatusCode};
use std::sync::{Arc, Mutex};
use std::time::Duration;

#[derive(Default)]
struct IdentityLoader {
    requests: Mutex<Vec<ResourceRequest>>,
}
#[async_trait]
impl ResourceLoader for IdentityLoader {
    async fn fetch(&self, request: ResourceRequest) -> Result<ResourceResponse, NetworkError> {
        let url = request.url.clone();
        self.requests.lock().unwrap().push(request);
        let mut headers = HeaderMap::new();
        headers.insert(
            http::header::CONTENT_TYPE,
            HeaderValue::from_static("text/html"),
        );
        Ok(ResourceResponse {
            status: StatusCode::OK,
            headers: headers.into(),
            body: b"<!doctype html><title>Persona</title>".to_vec(),
            effective_url: url,
            metadata: brimp_lite_worker::network::ResponseMetadata::default(),
        })
    }
}

#[test]
fn request_and_javascript_observe_one_coherent_identity() {
    let loader = Arc::new(IdentityLoader::default());
    let mut value =
        serde_json::to_value(brimp_lite_worker::persona::PersonaConfig::default()).unwrap();
    for (key, replacement) in serde_json::json!({
        "user_agent": "Mozilla/PersonaBrowser/150",
        "accept": "text/persona",
        "accept_language": "fr-CA,fr;q=0.9",
        "accept_encoding": "gzip, br",
        "sec_ch_ua": "\"Persona Chrome\";v=\"150\"",
        "sec_ch_ua_mobile": "?0",
        "sec_ch_ua_platform": "\"PersonaOS\"",
        "sec_ch_ua_arch": "\"arm\"",
        "sec_ch_ua_bitness": "\"32\"",
        "sec_ch_ua_model": "\"Desktop\"",
        "sec_ch_ua_platform_version": "\"27.1\"",
        "sec_ch_ua_full_version": "\"150.2.3.4\"",
        "sec_ch_ua_full_version_list": "\"Persona Chrome\";v=\"150.2.3.4\""
    })
    .as_object()
    .unwrap()
    {
        value["base_headers"][key] = replacement.clone();
    }
    value["viewport"] = serde_json::json!({"width":900,"height":700,"device_scale_factor":2});
    for (key, replacement) in serde_json::json!({
        "width":1800,"height":1200,"avail_width":1700,"avail_height":1100,
        "avail_left":20,"avail_top":30,"left":5,"top":6,
        "orientation_type":"landscape-secondary","orientation_angle":180
    })
    .as_object()
    .unwrap()
    {
        value["screen"][key] = replacement.clone();
    }
    value["window"] =
        serde_json::json!({"outer_width":920,"outer_height":760,"screen_x":40,"screen_y":50});
    for (key, replacement) in serde_json::json!({
        "platform":"PersonaPlatform","hardware_concurrency":12,"device_memory_gb":16,
        "max_touch_points":3,"connection_rtt_ms":42,"connection_downlink_mbps":"12.5",
        "connection_save_data":true,"do_not_track":"1","vendor":"Persona Vendor",
        "product_sub":"20260827","pdf_viewer_enabled":false
    })
    .as_object()
    .unwrap()
    {
        value["navigator"][key] = replacement.clone();
    }
    value["locale"] = serde_json::json!("fr-CA");
    value["languages"] = serde_json::json!(["fr-CA", "fr"]);
    let persona = brimp_lite_worker::persona::PersonaConfig::from_json(&value.to_string()).unwrap();
    let browser = Browser::with_persona_and_resource_loader(persona, loader.clone()).unwrap();
    let page = browser.new_page(PageOptions::default()).unwrap();
    page.navigate("https://identity.test/", Duration::from_secs(1))
        .unwrap();
    let observed = page.evaluate("({ userAgent: navigator.userAgent, appVersion: navigator.appVersion, platform: navigator.platform, language: navigator.language, languages: navigator.languages, hardwareConcurrency: navigator.hardwareConcurrency, deviceMemory: navigator.deviceMemory, maxTouchPoints: navigator.maxTouchPoints, connection: { type: navigator.connection.type, rtt: navigator.connection.rtt, downlink: navigator.connection.downlink, effectiveType: navigator.connection.effectiveType, saveData: navigator.connection.saveData }, doNotTrack: navigator.doNotTrack, vendor: navigator.vendor, productSub: navigator.productSub, pdfViewerEnabled: navigator.pdfViewerEnabled, webdriver: navigator.webdriver, uaData: [navigator.userAgentData.platform, navigator.userAgentData.brands[0].brand], viewport: [window.innerWidth, window.innerHeight, window.devicePixelRatio], screen: [screen.width, screen.height, screen.availWidth, screen.availHeight, screen.availLeft, screen.availTop, screen.colorDepth, screen.pixelDepth, screen.orientation.type, screen.orientation.angle], window: [window.outerWidth, window.outerHeight, window.screenX, window.screenY], screenPosition: [screen.left, screen.top], plugins: navigator.plugins.length })").unwrap();
    let request = &loader.requests.lock().unwrap()[0];
    assert_eq!(
        request
            .headers
            .get(http::header::USER_AGENT)
            .unwrap()
            .to_str()
            .unwrap(),
        observed["userAgent"]
    );
    for (name, expected) in [
        ("accept", "text/persona"),
        ("accept-language", "fr-CA,fr;q=0.9"),
        ("accept-encoding", "gzip, br"),
        ("sec-ch-ua", "\"Persona Chrome\";v=\"150\""),
        ("sec-ch-ua-mobile", "?0"),
        ("sec-ch-ua-platform", "\"PersonaOS\""),
        ("sec-ch-ua-arch", "\"arm\""),
    ] {
        assert_eq!(request.headers[name].to_str().unwrap(), expected);
    }
    assert_eq!(observed["appVersion"], "PersonaBrowser/150");
    assert_eq!(observed["platform"], "PersonaPlatform");
    assert_eq!(observed["language"], "fr-CA");
    assert_eq!(observed["languages"], serde_json::json!(["fr-CA", "fr"]));
    assert_eq!(observed["hardwareConcurrency"], 12);
    assert_eq!(observed["deviceMemory"], 16);
    assert_eq!(observed["maxTouchPoints"], 3);
    assert_eq!(
        observed["connection"],
        serde_json::json!({
            "type": "wifi",
            "rtt": 42,
            "downlink": 12.5,
            "effectiveType": "4g",
            "saveData": true
        })
    );
    assert_eq!(observed["doNotTrack"], "1");
    assert_eq!(observed["vendor"], "Persona Vendor");
    assert_eq!(observed["productSub"], "20260827");
    assert_eq!(observed["pdfViewerEnabled"], false);
    assert_eq!(observed["webdriver"], false);
    assert_eq!(
        observed["uaData"],
        serde_json::json!(["PersonaOS", "Persona Chrome"])
    );
    assert_eq!(observed["viewport"], serde_json::json!([900, 700, 2]));
    assert_eq!(
        observed["screen"],
        serde_json::json!([
            1800,
            1200,
            1700,
            1100,
            20,
            30,
            30,
            30,
            "landscape-secondary",
            180
        ])
    );
    assert_eq!(observed["window"], serde_json::json!([920, 760, 40, 50]));
    assert_eq!(observed["plugins"], 0);
    assert_eq!(observed["screenPosition"], serde_json::json!([5, 6]));
    let hints = page.evaluate_remote("navigator.userAgentData.getHighEntropyValues(['architecture','bitness','model','platformVersion','uaFullVersion','fullVersionList'])", true, None, true).unwrap()["value"].clone();
    assert_eq!(
        hints,
        serde_json::json!({
            "brands":[{"brand":"Persona Chrome","version":"150"}], "mobile":false,"platform":"PersonaOS",
            "architecture":"arm","bitness":"32","model":"Desktop","platformVersion":"27.1",
            "uaFullVersion":"150.2.3.4","fullVersionList":[{"brand":"Persona Chrome","version":"150.2.3.4"}]
        })
    );
}

#[test]
fn browser_builtins_have_native_sources_and_web_idl_shapes() {
    let loader = Arc::new(IdentityLoader::default());
    let browser = Browser::with_persona_and_resource_loader(
        brimp_lite_worker::persona::PersonaConfig::default(),
        loader,
    )
    .unwrap();
    let page = browser.new_page(PageOptions::default()).unwrap();
    let observed = page
        .evaluate(
            r#"(() => {
                function authoredFunction() { return 42; }
                const userAgent = Object.getOwnPropertyDescriptor(
                    Navigator.prototype,
                    "userAgent",
                );
                const querySelector = Document.prototype.querySelector;
                const nonNativeBuiltins = [];
                for (const name of Object.getOwnPropertyNames(globalThis)) {
                    const constructor = globalThis[name];
                    if (!/^[A-Z]/.test(name) || typeof constructor !== "function") continue;
                    const constructorSource = Function.prototype.toString.call(constructor);
                    if (!constructorSource.includes("[native code]")) {
                        nonNativeBuiltins.push(name);
                    }
                    const prototype = constructor.prototype;
                    if (!prototype) continue;
                    for (const key of Reflect.ownKeys(prototype)) {
                        const descriptor = Object.getOwnPropertyDescriptor(prototype, key);
                        for (const [kind, fn] of [
                            ["value", descriptor.value],
                            ["get", descriptor.get],
                            ["set", descriptor.set],
                        ]) {
                            if (
                                typeof fn === "function"
                                && !Function.prototype.toString.call(fn).includes("[native code]")
                            ) nonNativeBuiltins.push(`${name}.${String(key)}:${kind}`);
                        }
                    }
                }
                return {
                    navigatorHasOwnUserAgent: Object.hasOwn(navigator, "userAgent"),
                    navigatorGetter: Function.prototype.toString.call(userAgent.get),
                    navigatorDescriptor: {
                        enumerable: userAgent.enumerable,
                        configurable: userAgent.configurable,
                        hasSetter: typeof userAgent.set === "function",
                    },
                    documentConstructor: Function.prototype.toString.call(Document),
                    querySelector: Function.prototype.toString.call(querySelector),
                    functionToString: Function.prototype.toString.call(Function.prototype.toString),
                    authoredFunction: Function.prototype.toString.call(authoredFunction),
                    timeout: Function.prototype.toString.call(setTimeout),
                    uaDataMethod: Function.prototype.toString.call(
                        NavigatorUAData.prototype.getHighEntropyValues,
                    ),
                    screenTag: Object.prototype.toString.call(screen),
                    pluginsTag: Object.prototype.toString.call(navigator.plugins),
                    pluginsPrototype: Object.getPrototypeOf(navigator.plugins) === PluginArray.prototype,
                    pluginsIsArray: Array.isArray(navigator.plugins),
                    pluginMimeRelationship: navigator.plugins.length === 0
                        || navigator.plugins[0][0].enabledPlugin === navigator.plugins[0],
                    uaBrands: navigator.userAgentData.brands,
                    doNotTrack: navigator.doNotTrack,
                    coherentScreen: screen.width >= innerWidth
                        && screen.height >= innerHeight
                        && screen.availWidth <= screen.width
                        && screen.availHeight <= screen.height
                        && screen.colorDepth === screen.pixelDepth,
                    constructorDescriptors: [Document, PluginArray, Screen].map(constructor => {
                        const descriptor = Object.getOwnPropertyDescriptor(
                            globalThis,
                            constructor.name,
                        );
                        return [descriptor.writable, descriptor.enumerable, descriptor.configurable];
                    }),
                    markerLeaked: "__brimpMarkWebBuiltin" in globalThis,
                    hostBridgeLeaked: "__brimp" in globalThis,
                    bindingLexicalsLeaked: typeof __eventListeners !== "undefined",
                    nonNativeBuiltins,
                };
            })()"#,
        )
        .unwrap();

    assert_eq!(observed["navigatorHasOwnUserAgent"], false);
    assert_eq!(
        observed["navigatorGetter"],
        "function get userAgent() { [native code] }"
    );
    assert_eq!(
        observed["navigatorDescriptor"],
        serde_json::json!({
            "enumerable": true,
            "configurable": true,
            "hasSetter": false,
        })
    );
    assert_eq!(
        observed["documentConstructor"],
        "function Document() { [native code] }"
    );
    assert_eq!(
        observed["querySelector"],
        "function querySelector() { [native code] }"
    );
    assert_eq!(
        observed["functionToString"],
        "function toString() { [native code] }"
    );
    assert!(
        observed["authoredFunction"]
            .as_str()
            .unwrap()
            .contains("return 42")
    );
    assert_eq!(
        observed["timeout"],
        "function setTimeout() { [native code] }"
    );
    assert_eq!(
        observed["uaDataMethod"],
        "function getHighEntropyValues() { [native code] }"
    );
    assert_eq!(observed["screenTag"], "[object Screen]");
    assert_eq!(observed["pluginsTag"], "[object PluginArray]");
    assert_eq!(observed["pluginsPrototype"], true);
    assert_eq!(observed["pluginsIsArray"], false);
    assert_eq!(observed["pluginMimeRelationship"], true);
    assert_eq!(
        observed["uaBrands"],
        serde_json::json!([
            {"brand": "Not;A=Brand", "version": "8"},
            {"brand": "Chromium", "version": "150"},
            {"brand": "Google Chrome", "version": "150"},
        ])
    );
    assert_eq!(observed["doNotTrack"], serde_json::Value::Null);
    assert_eq!(observed["coherentScreen"], true);
    assert_eq!(
        observed["constructorDescriptors"],
        serde_json::json!([
            [true, false, true],
            [true, false, true],
            [true, false, true]
        ])
    );
    assert_eq!(observed["markerLeaked"], false);
    assert_eq!(observed["hostBridgeLeaked"], false);
    assert_eq!(observed["bindingLexicalsLeaked"], false);
    assert_eq!(observed["nonNativeBuiltins"], serde_json::json!([]));
}

#[test]
fn navigator_flags_control_supported_apis() {
    let mut persona = brimp_lite_worker::persona::PersonaConfig::default();
    persona.navigator.permissions_enabled = false;
    persona.navigator.bluetooth_enabled = false;
    persona.navigator.media_devices_enabled = false;
    persona.navigator.offscreen_canvas_enabled = false;
    persona.navigator.expose_global_privacy_control = true;
    persona.navigator.global_privacy_control = true;
    let browser =
        Browser::with_persona_and_resource_loader(persona, Arc::new(IdentityLoader::default()))
            .unwrap();
    let page = browser.new_page(PageOptions::default()).unwrap();
    assert_eq!(page.evaluate("({ permissions: 'permissions' in navigator, bluetooth: 'bluetooth' in navigator, mediaDevices: 'mediaDevices' in navigator, offscreen: typeof OffscreenCanvas, transfer: typeof HTMLCanvasElement.prototype.transferControlToOffscreen, privacy: navigator.globalPrivacyControl, canvas: typeof document.createElement('canvas').getContext })").unwrap(), serde_json::json!({
        "permissions":false,"bluetooth":false,"mediaDevices":false,"offscreen":"undefined","transfer":"undefined","privacy":true,"canvas":"function"
    }));
}

#[test]
fn notification_permission_and_bluetooth_availability_follow_persona() {
    let mut persona = brimp_lite_worker::persona::PersonaConfig::default();
    persona.navigator.notification_permission = "denied".into();
    persona.navigator.bluetooth_available = false;
    let browser =
        Browser::with_persona_and_resource_loader(persona, Arc::new(IdentityLoader::default()))
            .unwrap();
    let page = browser.new_page(PageOptions::default()).unwrap();
    assert_eq!(page.evaluate_remote("Promise.all([Notification.requestPermission(), navigator.permissions.query({ name: 'notifications' }).then(p => p.state), navigator.bluetooth.getAvailability(), navigator.mediaDevices.enumerateDevices()])", true, None, true).unwrap()["value"], serde_json::json!(["denied","denied",false,[]]));
}

#[test]
fn example_without_client_hints_omits_them_from_requests_and_navigator() {
    let persona = brimp_lite_worker::persona::PersonaConfig::from_json(include_str!(
        "../../../../examples/cli/safari-persona.json"
    ))
    .unwrap();
    let loader = Arc::new(IdentityLoader::default());
    let browser = Browser::with_persona_and_resource_loader(persona, loader.clone()).unwrap();
    let page = browser.new_page(PageOptions::default()).unwrap();
    page.navigate("https://identity.test/", Duration::from_secs(1))
        .unwrap();
    assert_eq!(
        page.evaluate("'userAgentData' in navigator").unwrap(),
        false
    );
    let requests = loader.requests.lock().unwrap();
    for (name, _) in requests[0].headers.iter() {
        assert!(!name.as_str().starts_with("sec-ch-ua"));
    }
}
